dm-agent/src/main.rs

```diff
@@ -0,0 +1,190 @@
+use futures::future::FutureExt;
+use mqtt_client::{Client, Message, MessageStream, QoS, Topic, TopicFilter};
+use std::str::from_utf8;
+use std::sync::Arc;
+use thin_edge_json::{
+    group::MeasurementGrouper, measurement::current_timestamp, measurement::FlatMeasurementVisitor,
+    serialize::ThinEdgeJsonSerializer,
+};
+use tokio::select;
+use tokio::time::Duration;
+use tracing::{debug_span, error, info, instrument, Instrument};
+
+const DEFAULT_STATS_COLLECTION_INTERVAL: u64 = 1000;
+
+const DEFAULT_HOST: &str = "localhost";
+const DEFAULT_PORT: u16 = 1883;
+const CLIENT_ID: &str = "tedge-dm-agent";
+const SOURCE_TOPIC: &str = "collectd/#";
+const TARGET_TOPIC: &str = "tedge/measurements";
+
+const APP_NAME: &str = "tedge-dm-agent";
+const DEFAULT_LOG_LEVEL: &str = "warn";
+const TIME_FORMAT: &str = "%Y-%m-%dT%H:%M:%S%.3f%:z";
+
+#[tokio::main]
+async fn main() {
+    let filter = std::env::var("RUST_LOG").unwrap_or_else(|_| DEFAULT_LOG_LEVEL.into());
+    tracing_subscriber::fmt()
+        .with_timer(tracing_subscriber::fmt::time::ChronoUtc::with_format(
+            TIME_FORMAT.into(),
+        ))
+        .with_env_filter(filter)
+        .with_span_events(tracing_subscriber::fmt::format::FmtSpan::CLOSE)
+        .init();
+
+    info!("{} starting!", APP_NAME);
+
+    if let Err(err) = run().instrument(debug_span!(APP_NAME)).await {
+        error!("{}", err);
```

Comment:
What I really wanted was to avoid this dm-agent from crashing even if it receives an invalid message. But now I realise that this isn't enough. What I need to do is to prevent the message processing select loop from crashing, but allow other unrecoverable errors to propagate up.